File: strategy/marvin/test/move/keepertest.lua
local KeeperTest = Class("Test.Move.KeeperTest", require "group/move/base")

local DebugCommands = require "../base/debugcommands"
local World = require "../base/world"
local Goal = require "observer/goal"
local Halt = require "task/shared/halt"
local Keeper = require "task/keeper/keeper"
local HallucinatingKeeper = require "task/test/hallucinatingkeeper"
local IO = require "util/io"
local G = World.Geometry

KeeperTest.MIN_ROBOTS = 1
KeeperTest.MAX_ROBOTS = 1

local SHOOT_SPEED = 6.5

local MIN_DISTANCE = 3.8
local MAX_DISTANCE = 3.8
local DISTANCE_INCREMENT = 0

local MIN_ANGLE = 1/4 * math.pi
local MAX_ANGLE = 3/4 * math.pi
local ANGLE_INCREMENT = 1/12 * math.pi

local RECORD = false
local FILENAME = "crescent"
local DESTINATION = "test/move/balldata/"..FILENAME..".balldata"

function KeeperTest.canStart()
	return true
end

function KeeperTest:_init()
	self._startTime = World.Time
	self._state = "Prepare"
	self._distance = MIN_DISTANCE
	self._angle = MIN_ANGLE
	self._shootLeft = false
	if RECORD then
		IO.save(DESTINATION, {})
	end
end

function KeeperTest:_canContinue()
	return true
end

function KeeperTest:_increment()
	self._shootLeft = not self._shootLeft

	if self._shootLeft == false then
		self._angle = self._angle + ANGLE_INCREMENT
		if self._angle > MAX_ANGLE then
			self._angle = MIN_ANGLE
		end

		self._distance = self._distance + DISTANCE_INCREMENT
		if self._distance > MAX_DISTANCE then
			self._distance = MIN_DISTANCE
		end

		local angle = self._angle / math.pi
		local message = "New Shot from distance "..tostring(self._distance).." and angle "..tostring(angle)
		log(message)
		if RECORD then
			IO.append(DESTINATION, message)
		end
	end
end

function KeeperTest:_update()
	local goal = G.FriendlyGoal
	local startPos = goal + Vector.fromAngle(self._angle):setLength(self._distance)

	-- append
	if RECORD then
		local speedVector = World.Ball.speed
		local spdX = speedVector.x
		local spdY = speedVector.y
		local relativePos = (World.Ball.pos - goal)
		local relX = relativePos.x
		local relY = relativePos.y

		local atkPos, atkDir, isShot = Goal.predictShot()
		IO.append(DESTINATION, tostring(relX).." "..tostring(relY).." "..tostring(spdX)
				.." "..tostring(spdY).." "..tostring(atkPos.x).." "..tostring(atkPos.y)
				.." "..tostring(atkDir.x).." "..tostring(atkDir.y).." "..tostring(isShot))
	end

	if self._state == "Prepare" and World.Time - self._startTime > 2 then
		self._state = "Shot"
		self._startTime = World.Time
		local targetPos = self._shootLeft and Vector(goal.x + 0.5, goal.y) or Vector(goal.x - 0.5, goal.y)
		local ball = {
			pos = startPos,
			posZ = 0,
			speedZ = 0,
			speed = (targetPos - startPos):setLength(SHOOT_SPEED) -- shoot with max speed
		}
		DebugCommands.moveObjects(ball)
		self:_increment()
	elseif self._state == "Shot" and World.Time - self._startTime > 3 then
		self._state = "Prepare"
		self._startTime = World.Time
		local ball = {
			pos = startPos,
			posZ = 0,
			speedZ = 0,
			speed = Vector(0, 0)
		}
		if World.IsSimulated then
			DebugCommands.moveObjects(ball)
		end
	end
end


function KeeperTest:_updateTasks()
	local taskAssignments = {}

	if World.IsSimulated then
		self:_update()
		taskAssignments[self._robots[1]] = {class = Keeper, params = {}, restart = false}
	elseif not RECORD then
		taskAssignments[self._robots[1]] = {class = HallucinatingKeeper, params = {DESTINATION}}
	else
		taskAssignments[self._robots[1]] = {class = Halt, params = {}}
	end

	return taskAssignments, self._robots[1]
end

return KeeperTest

File: strategy/typescript/glados/test/move/keepertest.ts
let KeeperTest = Class("Test.Move.KeeperTest", require "group/move/base")

let DebugCommands = require "+/base/debugcommands"
import * as World from "base/world";
import * as Goal from "glados/observer/goal";
let Halt = require "task/shared/halt"
let Keeper = require "task/keeper/keeper"
let HallucinatingKeeper = require "task/test/hallucinatingkeeper"
let IO = require "util/io"
let G = World.Geometry

KeeperTest.MIN_ROBOTS = 1
KeeperTest.MAX_ROBOTS = 1

let SHOOT_SPEED = 6.5

let MIN_ANGLE = 1/4 * Math.PI
let MAX_ANGLE = 3/4 * Math.PI
let ANGLE_INCREMENT = 1/12 * Math.PI

let MIN_DISTANCE = 3.8
let MAX_DISTANCE = 3.8
let DISTANCE_INCREMENT = 0

let RECORD = false
let FILENAME = "crescent"
let DESTINATION = "test/move/balldata/"+FILENAME+".balldata"

let HALLUCINATE_SIMULATOR = false
let HALT = true


// Instructions:

// Using a preexisting balldata file:
// move a .balldata file you want to use to the folder "marvin/test/move/balldata"
// .balldata files can be found on the NAS or recorded manually (more on that later)
// change the FILENAME constant to the name of the .balldata file
// On an actual field the robot will proceed to chase the imaginary ball
// If you want to have to robot hallucinate even in the simulator, set the HALLUCINATE_SIMULATOR flag to true

// Recording a balldata file:
// Specify a filename in the FILENAME constant and set the RECORD flag to true
// NOTE: If the name already exists it will be overwritten
// The move will record as long as the strategy is running and the RECORD flag is true,
// the test will run on repeat indefinitely

// Specifying test shots:
// Shots will always alternate between the left and right side of the goal
// After both sides of the goal have been hit, the angle, starting from MIN_ANGLE, will increment by ANGLE_INCREMENT
// Should the new angle then exceed the MAX_ANGLE, it will be reset to MIN_ANGLE and the distance will be incremented
// Distance will start at MIN_DISTANCE and increment by DISTANCE_INCREMENT
// Should the new distance exceed the MAX_DISTANCE, it will be reset to MIN_DISTANCE
// Shoot speed can be specified in the constant SHOOT_SPEED

// Visualisations:
// the visualisation "test/move/keepertest: Imaginary Ball" will display the ball from the .balldata file
// the visualisation "test/move/keepertest: Hit" will put a red marker on the keeper if it has touched the ball
// This is to evaluate how centrally the ball would have been caught
// This marker will be reset every shot

function KeeperTest.canStart () {
	return true
}

function KeeperTest:_init () {
	this._startTime = World.Time
	this._state = "Prepare"
	this._distance = MIN_DISTANCE
	this._angle = MIN_ANGLE
	this._shootLeft = false
	if (RECORD) {
		IO.save(DESTINATION, {})
	}
}

function KeeperTest:_canContinue () {
	return true
}

function KeeperTest:_increment () {
	this._shootLeft = not this._shootLeft

	if (this._shootLeft == false) {
		this._angle = this._angle + ANGLE_INCREMENT
		if (this._angle > MAX_ANGLE) {
			this._angle = MIN_ANGLE
		}

		this._distance = this._distance + DISTANCE_INCREMENT
		if (this._distance > MAX_DISTANCE) {
			this._distance = MIN_DISTANCE
		}
	}

	let angle = this._angle / Math.PI
	let message = "New Shot from distance "+String(this._distance)+" && angle "+String(angle)
	log(message)
	if (RECORD) {
		IO.append(DESTINATION, message)
	}
}

function KeeperTest:_update () {
	let goal = G.FriendlyGoal
	let startPos = goal + Vector.fromAngle(this._angle).setLength(this._distance)

	// append
	if (RECORD) {
		let speedVector = World.Ball.speed
		let spdX = speedVector.x
		let spdY = speedVector.y
		let relativePos = (World.Ball.pos - goal)
		let relX = relativePos.x
		let relY = relativePos.y

		let atkPos, atkDir, isShot = Goal.predictShot()
		IO.append(DESTINATION, String(relX)+" "+String(relY)+" "+String(spdX)
				 + " "+String(spdY)+" "+String(atkPos.x)+" "+String(atkPos.y)
				 + " "+String(atkDir.x)+" "+String(atkDir.y)+" "+String(isShot))
	}

	if (this._state == "Prepare" && World.Time - this._startTime > 2) {
		this._state = "Shot"
		this._startTime = World.Time
		let targetPos = this._shootLeft ? Vector(goal.x + 0.5, goal.y) : Vector(goal.x - 0.5, goal.y)
		let ball = {
			pos = startPos,
			posZ = 0,
			speedZ = 0,
			speed = (targetPos - startPos).setLength(SHOOT_SPEED) // shoot with max speed
		}
		DebugCommands.moveObjects(ball)
		this._increment()
	} else if (this._state == "Shot" && World.Time - this._startTime > 3) {
		this._state = "Prepare"
		this._startTime = World.Time
		let ball = {
			pos = startPos,
			posZ = 0,
			speedZ = 0,
			speed = new Vector(0, 0)
		}
		if (World.IsSimulated) {
			DebugCommands.moveObjects(ball)
		}
	}
}


function KeeperTest:_updateTasks () {
	let taskAssignments = {}

	if (World.IsSimulated && not HALLUCINATE_SIMULATOR) {
		this._update()
		taskAssignments[this._robots[0]] = {class: HALT ? Halt : Keeper, params: {}, restart: false}
	} else if (not RECORD) {
		taskAssignments[this._robots[0]] = {class: HallucinatingKeeper, params: {DESTINATION}}
	} else {
		taskAssignments[this._robots[0]] = {class: Halt, params: {}}
	}

	return taskAssignments, this._robots[0]
}

return KeeperTest